File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string s){
        int l = 0, r = s.length()-1;

        while (l<r){
            if (s[l]!=s[r]) return false;
            l++;r--;
        }

        return true;
    }
    vector<vector<string>> partition(string s) {
    int n = s.length();
    vector<vector<vector<string>>> dp(n + 1);

    dp[0] = {{}};

    for (int i = 1; i <= n; i++) {
        for (int j = 0; j < i; j++) {
            string sub = s.substr(j, i - j);
            if (isPalindrome(sub)) {
                for (auto &v : dp[j]) {
                    vector<string> newPartition = v;
                    newPartition.push_back(sub);
                    dp[i].push_back(newPartition);
                }
            }
        }
    }

    return dp[n];
}
};

File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []

        def is_pal(sub):
            return sub == sub[::-1]

        def backtrack(start, path):
            if start == len(s):
                res.append(path[:])
                return
            
            for end in range(start + 1, len(s) + 1):
                sub = s[start:end]
                if is_pal(sub):
                    path.append(sub)
                    backtrack(end, path)
                    path.pop()

        backtrack(0, [])
        return res